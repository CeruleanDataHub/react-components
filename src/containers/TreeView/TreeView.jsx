import "react-sortable-tree/style.css";

import PropTypes from "prop-types";
import React from "react";
import SortableTree from "react-sortable-tree";
import styled from "styled-components";

const Container = styled.div`
  height: 100%;
`;

export const TreeView = ({
  treeData,
  onChange,
  canDrag,
  generateNodeProps
}) => (
  <Container>
    <SortableTree
      treeData={treeData}
      onChange={onChange}
      canDrag={canDrag}
      generateNodeProps={generateNodeProps}
    />
  </Container>
);

TreeView.propTypes = {
  treeData: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string,
      subtitle: PropTypes.string,
      // eslint-disable-next-line react/forbid-prop-types
      children: PropTypes.array
    })
  ),
  onChange: PropTypes.func,
  canDrag: PropTypes.bool,
  generateNodeProps: PropTypes.func
};

TreeView.defaultProps = {
  treeData: [],
  onChange: () => null,
  canDrag: false,
  generateNodeProps: () => null
};
