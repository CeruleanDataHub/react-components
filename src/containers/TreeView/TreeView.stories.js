import React, { useRef, useState } from "react";

import { Button } from "../Button/Button";
import { Icon } from "../Icon/Icon";
import { Popover } from "../Popover/Popover";
import { TreeView } from "./TreeView";

export default {
  title: "TreeView",
  component: TreeView
};

const treeData = [
  {
    title: "Top level parent",
    expanded: true,
    subtitle: "Subtitle",
    children: [
      {
        title: "Parent A",
        children: [{ title: "Child 1" }, { title: "Child 2" }]
      },
      {
        title: "Parent B",
        children: [{ title: "Child 3" }, { title: "Child 4" }]
      },
      {
        title: "Parent C",
        children: [{ title: "Child 5" }, { title: "Child 6" }]
      }
    ]
  }
];

export const TreeViewStory = () => {
  const [data, setData] = useState(treeData);

  return (
    <div style={{ height: 500 }}>
      <TreeView treeData={data} onChange={tree => setData(tree)} />
    </div>
  );
};

export const TreeViewDraggableStory = () => {
  const [data, setData] = useState(treeData);

  return (
    <div style={{ height: 500 }}>
      <TreeView treeData={data} onChange={tree => setData(tree)} canDrag />
    </div>
  );
};

const AdditionalNodeProp = path => {
  const [popoverOpen, setPopoverOpen] = useState(false);
  const moreRef = useRef(null);

  return (
    <>
      <Button
        key={path}
        onClick={() => setPopoverOpen(!popoverOpen)}
        ref={moreRef}
      >
        <Icon name="more" />
      </Button>
      <Popover isOpen={popoverOpen} containerRef={moreRef}>
        <p>
          Path:&nbsp;
          {path.path}
        </p>
      </Popover>
    </>
  );
};

export const TreeViewAdditionalNodePropsStory = () => {
  const [data, setData] = useState(treeData);

  const handleGenerateNodeProps = ({ path }) => ({
    buttons: [<AdditionalNodeProp path={path} />]
  });

  return (
    <div style={{ height: 500 }}>
      <TreeView
        treeData={data}
        onChange={tree => setData(tree)}
        generateNodeProps={handleGenerateNodeProps}
      />
    </div>
  );
};
