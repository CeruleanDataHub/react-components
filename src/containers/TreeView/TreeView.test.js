import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import { TreeView } from "./TreeView";

const treeData = [
  {
    title: "Top level parent",
    expanded: true,
    subtitle: "Subtitle",
    children: [
      {
        title: "Parent A",
        children: [{ title: "Child 1" }, { title: "Child 2" }]
      },
      {
        title: "Parent B",
        children: [{ title: "Child 3" }, { title: "Child 4" }]
      },
      {
        title: "Parent C",
        children: [{ title: "Child 5" }, { title: "Child 6" }]
      }
    ]
  }
];

describe("TreeView", () => {
  it("should render", () => {
    const component = shallow(<TreeView treeData={treeData} />);
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should render with canDrag prop", () => {
    const component = shallow(<TreeView treeData={treeData} canDrag />);
    expect(toJson(component)).toMatchSnapshot();
  });

  it("calls onChange handler", () => {
    const handleChange = jest.fn();
    const component = shallow(
      <TreeView treeData={treeData} onChange={handleChange} />
    );
    expect(handleChange).not.toHaveBeenCalled();
    component
      .find("DragDropContext(SortableTreeWithoutDndContext)")
      .simulate("change");
    expect(handleChange).toHaveBeenCalled();
  });

  it("calls default onChange handler if no prop is passed", () => {
    const component = shallow(<TreeView treeData={treeData} />);
    component
      .find("DragDropContext(SortableTreeWithoutDndContext)")
      .prop("onChange")();
    expect(toJson(component)).toMatchSnapshot();
  });

  it("calls default generateNodeProps handler if no prop is passed", () => {
    const component = shallow(<TreeView treeData={treeData} />);
    component
      .find("DragDropContext(SortableTreeWithoutDndContext)")
      .prop("generateNodeProps")();
    expect(toJson(component)).toMatchSnapshot();
  });
});
