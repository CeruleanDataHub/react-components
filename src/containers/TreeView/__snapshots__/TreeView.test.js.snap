// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeView calls default generateNodeProps handler if no prop is passed 1`] = `
<styled.div>
  <SortableTree
    canDrag={false}
    generateNodeProps={[Function]}
    onChange={[Function]}
    treeData={
      Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "title": "Child 1",
                },
                Object {
                  "title": "Child 2",
                },
              ],
              "title": "Parent A",
            },
            Object {
              "children": Array [
                Object {
                  "title": "Child 3",
                },
                Object {
                  "title": "Child 4",
                },
              ],
              "title": "Parent B",
            },
            Object {
              "children": Array [
                Object {
                  "title": "Child 5",
                },
                Object {
                  "title": "Child 6",
                },
              ],
              "title": "Parent C",
            },
          ],
          "expanded": true,
          "subtitle": "Subtitle",
          "title": "Top level parent",
        },
      ]
    }
  />
</styled.div>
`;

exports[`TreeView calls default onChange handler if no prop is passed 1`] = `
<styled.div>
  <SortableTree
    canDrag={false}
    generateNodeProps={[Function]}
    onChange={[Function]}
    treeData={
      Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "title": "Child 1",
                },
                Object {
                  "title": "Child 2",
                },
              ],
              "title": "Parent A",
            },
            Object {
              "children": Array [
                Object {
                  "title": "Child 3",
                },
                Object {
                  "title": "Child 4",
                },
              ],
              "title": "Parent B",
            },
            Object {
              "children": Array [
                Object {
                  "title": "Child 5",
                },
                Object {
                  "title": "Child 6",
                },
              ],
              "title": "Parent C",
            },
          ],
          "expanded": true,
          "subtitle": "Subtitle",
          "title": "Top level parent",
        },
      ]
    }
  />
</styled.div>
`;

exports[`TreeView should render 1`] = `
<styled.div>
  <SortableTree
    canDrag={false}
    generateNodeProps={[Function]}
    onChange={[Function]}
    treeData={
      Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "title": "Child 1",
                },
                Object {
                  "title": "Child 2",
                },
              ],
              "title": "Parent A",
            },
            Object {
              "children": Array [
                Object {
                  "title": "Child 3",
                },
                Object {
                  "title": "Child 4",
                },
              ],
              "title": "Parent B",
            },
            Object {
              "children": Array [
                Object {
                  "title": "Child 5",
                },
                Object {
                  "title": "Child 6",
                },
              ],
              "title": "Parent C",
            },
          ],
          "expanded": true,
          "subtitle": "Subtitle",
          "title": "Top level parent",
        },
      ]
    }
  />
</styled.div>
`;

exports[`TreeView should render with canDrag prop 1`] = `
<styled.div>
  <SortableTree
    canDrag={true}
    generateNodeProps={[Function]}
    onChange={[Function]}
    treeData={
      Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "title": "Child 1",
                },
                Object {
                  "title": "Child 2",
                },
              ],
              "title": "Parent A",
            },
            Object {
              "children": Array [
                Object {
                  "title": "Child 3",
                },
                Object {
                  "title": "Child 4",
                },
              ],
              "title": "Parent B",
            },
            Object {
              "children": Array [
                Object {
                  "title": "Child 5",
                },
                Object {
                  "title": "Child 6",
                },
              ],
              "title": "Parent C",
            },
          ],
          "expanded": true,
          "subtitle": "Subtitle",
          "title": "Top level parent",
        },
      ]
    }
  />
</styled.div>
`;
